[MRG+1] Update class_weight.py

MMeketon opened the pull request:
closes #8312 
Changed ValueError message to reflect that a class label may be non-integer

#### Reference Issue #8312 

#### What does this implement/fix? Explain your changes.
The original line `ValueError("Class label %d not present." %c)` implicitly assumed the class label `c` was an integer.  The fix `ValueError("Class label {} not present.".format(c))` allows the class label to be a more general type, such as a string

#### Any other comments?
This is my first ever attempt to fix anything in "open source" using github.  Let me know if I goofed in the process.

MMeketon committed:
Update class_weight.py

Changed ValueError message to reflect that a class label may be non-integer

TomDLT reviewed (approved):
LGTM

jnothman commented:
thanks. please add a test

On 9 Feb 2017 6:02 am, "Tom Dupré la Tour" <notifications@github.com> wrote:

> *@TomDLT* approved this pull request.
>
> LGTM
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8319#pullrequestreview-20840217>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-QKumDztvjUt5cstLw96e1aLAYdks5rahFcgaJpZM4L7BOb>
> .
>

MMeketon commented:
Joel,

This is my first time with open source, GitHub, etc.  Can you point me to documentation on how others add a test, do a merge, etc?

From: Joel Nothman [mailto:notifications@github.com]
Sent: Wednesday, February 08, 2017 3:27 PM
To: scikit-learn/scikit-learn
Cc: Meketon, Marc; Author
Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py (#8319)

thanks. please add a test

On 9 Feb 2017 6:02 am, "Tom Dupré la Tour" <notifications@github.com<mailto:notifications@github.com>> wrote:

> *@TomDLT* approved this pull request.
>
> LGTM
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8319#pullrequestreview-20840217>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-QKumDztvjUt5cstLw96e1aLAYdks5rahFcgaJpZM4L7BOb>
> .
>


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-278451141>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMpYD0MipENn94E-vExg1DAcIyeNQ4izks5raiUTgaJpZM4L7BOb>.

________________________________
This e-mail and any attachments may be confidential or legally privileged. If you received this message in error or are not the intended recipient, you should destroy the e-mail message and any attachments or copies, and you are prohibited from retaining, distributing, disclosing or using any information contained herein. Please inform us of the erroneous delivery by return e-mail. Thank you for your cooperation.

jnothman commented:
Does looking at sklearn/utils/tests/test_class_weight.py give you some idea?

On 9 February 2017 at 07:30, MMeketon <notifications@github.com> wrote:

> Joel,
>
> This is my first time with open source, GitHub, etc. Can you point me to
> documentation on how others add a test, do a merge, etc?
>
> From: Joel Nothman [mailto:notifications@github.com]
> Sent: Wednesday, February 08, 2017 3:27 PM
> To: scikit-learn/scikit-learn
> Cc: Meketon, Marc; Author
> Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py
> (#8319)
>
> thanks. please add a test
>
> On 9 Feb 2017 6:02 am, "Tom Dupré la Tour" <notifications@github.com<
> mailto:notifications@github.com>> wrote:
>
> > *@TomDLT* approved this pull request.
> >
> > LGTM
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scikit-learn/scikit-learn/pull/8319#
> pullrequestreview-20840217>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAEz6-
> QKumDztvjUt5cstLw96e1aLAYdks5rahFcgaJpZM4L7BOb>
> > .
> >
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub<https://github.com/
> scikit-learn/scikit-learn/pull/8319#issuecomment-278451141>, or mute the
> thread<https://github.com/notifications/unsubscribe-auth/AMpYD0MipENn94E-
> vExg1DAcIyeNQ4izks5raiUTgaJpZM4L7BOb>.
>
> ________________________________
> This e-mail and any attachments may be confidential or legally privileged.
> If you received this message in error or are not the intended recipient,
> you should destroy the e-mail message and any attachments or copies, and
> you are prohibited from retaining, distributing, disclosing or using any
> information contained herein. Please inform us of the erroneous delivery by
> return e-mail. Thank you for your cooperation.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-278452140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66yLZIutvyhxF9Vj-sN3ZpVs9m7kks5raiXsgaJpZM4L7BOb>
> .
>

MMeketon commented:
Yes.  Very helpful.  Will create the test tonight.  Appreciate this pointer.

From: Joel Nothman [mailto:notifications@github.com]
Sent: Wednesday, February 08, 2017 7:01 PM
To: scikit-learn/scikit-learn
Cc: Meketon, Marc; Author
Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py (#8319)

Does looking at sklearn/utils/tests/test_class_weight.py give you some idea?

On 9 February 2017 at 07:30, MMeketon <notifications@github.com<mailto:notifications@github.com>> wrote:

> Joel,
>
> This is my first time with open source, GitHub, etc. Can you point me to
> documentation on how others add a test, do a merge, etc?
>
> From: Joel Nothman [mailto:notifications@github.com]
> Sent: Wednesday, February 08, 2017 3:27 PM
> To: scikit-learn/scikit-learn
> Cc: Meketon, Marc; Author
> Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py
> (#8319)
>
> thanks. please add a test
>
> On 9 Feb 2017 6:02 am, "Tom Dupré la Tour" <notifications@github.com<
<mailto:notifications@github.com%3c%0b>> mailto:notifications@github.com>> wrote:
>
> > *@TomDLT* approved this pull request.
> >
> > LGTM
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scikit-learn/scikit-learn/pull/8319#
> pullrequestreview-20840217>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAEz6-
<https://github.com/notifications/unsubscribe-auth/AAEz6-%0b>> QKumDztvjUt5cstLw96e1aLAYdks5rahFcgaJpZM4L7BOb>
> > .
> >
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub<https://github.com/
<https://github.com/%0b>> scikit-learn/scikit-learn/pull/8319#issuecomment-278451141>, or mute the
> thread<https://github.com/notifications/unsubscribe-auth/AMpYD0MipENn94E-
<https://github.com/notifications/unsubscribe-auth/AMpYD0MipENn94E-%0b>> vExg1DAcIyeNQ4izks5raiUTgaJpZM4L7BOb>.
>
> ________________________________
> This e-mail and any attachments may be confidential or legally privileged.
> If you received this message in error or are not the intended recipient,
> you should destroy the e-mail message and any attachments or copies, and
> you are prohibited from retaining, distributing, disclosing or using any
> information contained herein. Please inform us of the erroneous delivery by
> return e-mail. Thank you for your cooperation.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-278452140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66yLZIutvyhxF9Vj-sN3ZpVs9m7kks5raiXsgaJpZM4L7BOb>
> .
>


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-278502940>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMpYDwmbdyCChtM4iclASnSm0izoZUN_ks5ralcpgaJpZM4L7BOb>.

________________________________
This e-mail and any attachments may be confidential or legally privileged. If you received this message in error or are not the intended recipient, you should destroy the e-mail message and any attachments or copies, and you are prohibited from retaining, distributing, disclosing or using any information contained herein. Please inform us of the erroneous delivery by return e-mail. Thank you for your cooperation.

lesteve commented:
@MMeketon have you made progress on the test? It seems like a useful fix.

MMeketon commented:
Sorry, I got caught up on other work.  Will post the test later today.

From: Loïc Estève [mailto:notifications@github.com]
Sent: Wednesday, February 15, 2017 3:42 AM
To: scikit-learn/scikit-learn
Cc: Meketon, Marc; Mention
Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py (#8319)


@MMeketon<https://github.com/MMeketon> have you made progress on the test? It seems like a useful fix.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-279949856>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMpYD_70Ids4-B150tZE6kr4PLDLskC5ks5rcrpzgaJpZM4L7BOb>.

________________________________
This e-mail and any attachments may be confidential or legally privileged. If you received this message in error or are not the intended recipient, you should destroy the e-mail message and any attachments or copies, and you are prohibited from retaining, distributing, disclosing or using any information contained herein. Please inform us of the erroneous delivery by return e-mail. Thank you for your cooperation.

MMeketon commented:
I just submitted the test.

Being a newbie to this, it took awhile to establish the right test environment.  The test itself was pretty small.

-Marc

From: Joel Nothman [mailto:notifications@github.com]
Sent: Wednesday, February 08, 2017 7:01 PM
To: scikit-learn/scikit-learn
Cc: Meketon, Marc; Author
Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py (#8319)

Does looking at sklearn/utils/tests/test_class_weight.py give you some idea?

On 9 February 2017 at 07:30, MMeketon <notifications@github.com<mailto:notifications@github.com>> wrote:

> Joel,
>
> This is my first time with open source, GitHub, etc. Can you point me to
> documentation on how others add a test, do a merge, etc?
>
> From: Joel Nothman [mailto:notifications@github.com]
> Sent: Wednesday, February 08, 2017 3:27 PM
> To: scikit-learn/scikit-learn
> Cc: Meketon, Marc; Author
> Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py
> (#8319)
>
> thanks. please add a test
>
> On 9 Feb 2017 6:02 am, "Tom Dupré la Tour" <notifications@github.com<
<mailto:notifications@github.com%3c%0b>> mailto:notifications@github.com>> wrote:
>
> > *@TomDLT* approved this pull request.
> >
> > LGTM
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scikit-learn/scikit-learn/pull/8319#
> pullrequestreview-20840217>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAEz6-
<https://github.com/notifications/unsubscribe-auth/AAEz6-%0b>> QKumDztvjUt5cstLw96e1aLAYdks5rahFcgaJpZM4L7BOb>
> > .
> >
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub<https://github.com/
<https://github.com/%0b>> scikit-learn/scikit-learn/pull/8319#issuecomment-278451141>, or mute the
> thread<https://github.com/notifications/unsubscribe-auth/AMpYD0MipENn94E-
<https://github.com/notifications/unsubscribe-auth/AMpYD0MipENn94E-%0b>> vExg1DAcIyeNQ4izks5raiUTgaJpZM4L7BOb>.
>
> ________________________________
> This e-mail and any attachments may be confidential or legally privileged.
> If you received this message in error or are not the intended recipient,
> you should destroy the e-mail message and any attachments or copies, and
> you are prohibited from retaining, distributing, disclosing or using any
> information contained herein. Please inform us of the erroneous delivery by
> return e-mail. Thank you for your cooperation.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-278452140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66yLZIutvyhxF9Vj-sN3ZpVs9m7kks5raiXsgaJpZM4L7BOb>
> .
>


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-278502940>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMpYDwmbdyCChtM4iclASnSm0izoZUN_ks5ralcpgaJpZM4L7BOb>.

________________________________
This e-mail and any attachments may be confidential or legally privileged. If you received this message in error or are not the intended recipient, you should destroy the e-mail message and any attachments or copies, and you are prohibited from retaining, distributing, disclosing or using any information contained herein. Please inform us of the erroneous delivery by return e-mail. Thank you for your cooperation.

MMeketon commented:
I just added the test.

From: Loïc Estève [mailto:notifications@github.com]
Sent: Wednesday, February 15, 2017 3:42 AM
To: scikit-learn/scikit-learn
Cc: Meketon, Marc; Mention
Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py (#8319)


@MMeketon<https://github.com/MMeketon> have you made progress on the test? It seems like a useful fix.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-279949856>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMpYD_70Ids4-B150tZE6kr4PLDLskC5ks5rcrpzgaJpZM4L7BOb>.

________________________________
This e-mail and any attachments may be confidential or legally privileged. If you received this message in error or are not the intended recipient, you should destroy the e-mail message and any attachments or copies, and you are prohibited from retaining, distributing, disclosing or using any information contained herein. Please inform us of the erroneous delivery by return e-mail. Thank you for your cooperation.

lesteve commented:
> I just added the test.

It is not part of the diff. You probably haven't pushed your commit to your branch then.

MMeketon commented:
I think I figured it out.  Let me know if the “diff” is now there.



From: Loïc Estève [mailto:notifications@github.com]
Sent: Thursday, February 16, 2017 4:45 PM
To: scikit-learn/scikit-learn
Cc: Meketon, Marc; Mention
Subject: Re: [scikit-learn/scikit-learn] [MRG+1] Update class_weight.py (#8319)


I just added the test.

It is not part of the diff. You probably haven't pushed your commit to your branch then.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/8319#issuecomment-280470943>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMpYD-bldrtR-KfXUXPOZQHdZrfj-nzZks5rdMNGgaJpZM4L7BOb>.

________________________________
This e-mail and any attachments may be confidential or legally privileged. If you received this message in error or are not the intended recipient, you should destroy the e-mail message and any attachments or copies, and you are prohibited from retaining, distributing, disclosing or using any information contained herein. Please inform us of the erroneous delivery by return e-mail. Thank you for your cooperation.

jnothman commented:
No, nothing's changed.

lesteve commented:
You can look at the diff yourself clicking on the "Files changed" tab, which you can access through this URL:
https://github.com/scikit-learn/scikit-learn/pull/8319/files

MMeketon committed:
Test that the ValueError is raised an extra, string class label is found in the class_weights dictionary

Previously when class_weights were used, a check was made to see if
there were elements in the class_weights dictionary that were not in the
classes array.  If so, a ValueError was raised.  But there was a bug in
the formatting of the ValueError that assumed the classes were integer
and the ValueErorr was not raised when the classes were not numbers.
This test is to verify that the ValueError is properly raised.

lesteve commented on sklearn/utils/tests/test_class_weight.py:
```diff
@@ -117,6 +117,15 @@ def test_compute_class_weight_balanced_unordered():
+    classes = np.unique(y)
+    class_weights = {c: 1.0 for c in classes}
+    class_weights['D'] = 1.0  # This should get a proper ValueError
+    cw = assert_raises(ValueError, compute_class_weight, class_weights, 
```
Well done for going through your git/github initiation! There are a few ways you can simplify this test. One is to 
reuse `test_compute_class_weight_not_present` and add your test there rather than creating a new function after this [line](https://github.com/MMeketon/scikit-learn/blob/1dec75c18904b35ec147b7e1a0d9dbd80b089fbe/sklearn/utils/tests/test_class_weight.py#L33).

I think that just adding this should be enough:
```py
# Fix exception in error message formatting when missing label is a string
# https://github.com/scikit-learn/scikit-learn/issues/8312
assert_raise_message(ValueError,
                     'Class label label_not_present not present',
                     compute_class_weight,
                     {'label_not_present': 1.}, classes, y)
```

MMeketon replied:
I used `test_compute_class_weight_not_present` as a template for the test I introduced, but did not want to change it.  I'm personally OK with multiple asserts in a single test case, but others (e.g., Robert Martin, "Clean Code", Ch 9) are against it, so I didn't want to upset others with multiple asserts.

I did incorporate your suggestion of using `assert_raise_message` which is an excellent idea - thank you.

MMeketon committed:
Recoded test to test for ValueError and the proper message being displayed

A big thank you to Loïc Estève for pointing out that an
assert_raises_message is a better test than assert_raises.

MMeketon reviewed (commented):
Made changes to make travis happier, and changed assert_raise to assert_raise_message.

lesteve committed:
Simplify test

lesteve commented:
I pushed the change I mentioned in https://github.com/scikit-learn/scikit-learn/pull/8319#discussion_r102061696. I'll wait for the CIs to be green and merge this one. @MMeketon please ping me if I forget to do it.

lesteve commented:
Merged, thanks a lot @MMeketon!
